src/main/java/dev/blaauwendraad/masker/json/KeyMatcher.java

```diff
@@ -178,14 +176,73 @@ KeyMaskingConfig getMaskConfigIfMatched(byte[] bytes, int keyOffset, int keyLeng
 
     @Nullable
     private TrieNode searchNode(byte[] bytes, int offset, int length) {
-        if (!knownKeyLengthsInBytes[Math.min(length, 255)]) {
-            return null;
-        }
         TrieNode node = root;
 
         for (int i = offset; i < offset + length; i++) {
             byte b = bytes[i];
-            node = node.child(b);
+            if (b == '\\' && bytes[i + 1] == 'u' && i <= offset + length - 6) {
+                int valueStartIndex = i;
+                char unicodeHexBytesAsChar = Utf8Util.unicodeHexToChar(bytes, i + 2);
+                i += 5; // -1 to offset loop increment
+                if (unicodeHexBytesAsChar < 0x80) {
+                    // < 128 (in decimal) fits in 7 bits which is 1 byte of data in UTF-8
+                    node = node.child((byte) unicodeHexBytesAsChar);
+                } else if (unicodeHexBytesAsChar < 0x800) { // 2048 in decimal,
+                    // < 2048 (in decimal) fits in 11 bits which is 2 bytes of data in UTF-8
+                    node = node.child((byte) (0xc0 | (unicodeHexBytesAsChar >> 6)));
+                    if (node == null) {
+                        return null;
+                    }
+                    node = node.child((byte) (0x80 | (unicodeHexBytesAsChar & 0x3f)));
+                } else if (Character.isSurrogate(unicodeHexBytesAsChar)) {
+                    // decoding non-BMP characters in UTF-16 using a pair of high and low
+                    // surrogates which together form one unicode character.
+                    int codePoint = -1;
+                    if (Character.isHighSurrogate(unicodeHexBytesAsChar) // first surrogate must be the high surrogate
+                        && i <= offset + length - 6 /* -6 for all bytes of the byte encoded unicode character (\\u + 4 hex bytes) to prevent possible ArrayIndexOutOfBoundsExceptions */
+                        && bytes[i] == '\\' // the high surrogate must be followed by a low surrogate (starting with \\u)
+                        && bytes[i + 1] == 'u'
+                    ) {
+                        char lowSurrogate = Utf8Util.unicodeHexToChar(bytes, i + 2);
+                        if (Character.isLowSurrogate(lowSurrogate)) {
+                            codePoint = Character.toCodePoint(unicodeHexBytesAsChar, lowSurrogate);
+                        }
+                    }
+                    if (codePoint < 0) {
+                        // the key contains invalid surrogate pair and won't be matched
+                        return null;
```

Comment:
@Breus coming back to our discussion 😄 
Noticed that JSONTestSuite `i_object_key_lone_2nd_surrogate.json` started to fail as the key contains incomplete surrogate. Unlike in `withTextFunction` I think here it makes more sense to treat it as "key didn't match" (would still be masked in allow-mode).
For reference jackson parses that key as `�`

Reply:
Perhaps it's one more reason to adhere to jackson behavior in `withTextFunction` as well